File: hs/src/Q800/Q801.hs
module Q800.Q801
(
    minSwapPlanA,
    minSwapPlanB
) where

minSwapPlanA :: [Int] -> [Int] -> ([Int],[Int])
minSwapPlanA a b = plana a b
        where
            plana [] _ = ([],[])
            plana _ [] = ([],[])
            plana [x] y = ([x],y)
            plana x [y] = (x,[y])
            plana (x1:xs@(x2:_)) (y1:ys@(y2:_))
                | x1 >= x2 || y1 >= y2 = (\(nx,ny) -> (x1:nx, y1:ny)) (plana ys xs)
                | otherwise = (\(nx,ny) -> (x1:nx, y1:ny)) (plana ys xs)

minSwapPlanB :: [Int] -> [Int] -> Int
minSwapPlanB a b = planb a b 0 1
        where
            planb [] _ dp0 dp1 = min dp0 dp1
            planb _ [] dp0 dp1 = min dp0 dp1
            planb _ [_] dp0 dp1 = min dp0 dp1
            planb [_] _ dp0 dp1 = min dp0 dp1
            planb (x1:xs@(x2:_)) (y1:ys@(y2:_)) dp0 dp1
                | x1 < x2 && y1 < y2 && x1 < y2 && y1 < x2 = planb xs ys (min dp0 dp1) ((min dp0 dp1) + 1)
                | x1 < y2 && y1 < x2 = planb xs ys dp1 (dp0+1)
                | otherwise = planb xs ys dp0 (dp1 + 1)

File: golang/q800/q801.go
package q800

import (
	stupid_self "github.com/zoroqi/stupid-self/golang"
)

func minSwap(nums1 []int, nums2 []int) int {
	return minSwapPlanB(nums1, nums2)
}

func minSwapPlanA(nums1 []int, nums2 []int) ([]int, []int) {
	i := 0
	l := len(nums1) - 1
	for i < l {
		x1, x2 := nums1[i], nums1[i+1]
		y1, y2 := nums2[i], nums2[i+1]

		if x1 >= x2 || y1 >= y2 {
			nums1[i+1], nums2[i+1] = y2, x2
		}
		i++
	}
	return nums1, nums2
}

func IsSequences(numss ...[]int) bool {
	for _, nums := range numss {
		for i := 0; i < (len(nums) - 1); i++ {
			if nums[i] >= nums[i+1] {
				return false
			}
		}
	}
	return true
}

func minSwapPlanB(nums1 []int, nums2 []int) int {
	l := len(nums1)
	dp0, dp1 := 0, 1
	for i := 1; i < l; i++ {
		x1, x2 := nums1[i-1], nums1[i]
		y1, y2 := nums2[i-1], nums2[i]
		if x1 < x2 && y1 < y2 && x1 < y2 && y1 < x2 {
			dp0 = stupid_self.IntMin(dp0, dp1)
			dp1 = stupid_self.IntMin(dp0, dp1) + 1
		} else if x1 < y2 && y1 < x2 {
			dp0, dp1 = dp1, dp0+1
		} else {
			dp0, dp1 = dp0, dp1+1
		}
	}
	return stupid_self.IntMin(dp0, dp1)
}
